util/flightcontrol/flightcontrol.go

```diff
@@ -27,12 +27,29 @@ type Group struct {
 	m  map[string]*call // lazily initialized
 }
 
+type retries struct {
+	num     int
+	backoff time.Duration
+}
+
 // Do executes a context function syncronized by the key
 func (g *Group) Do(ctx context.Context, key string, fn func(ctx context.Context) (interface{}, error)) (v interface{}, err error) {
+	retries := retries{backoff: time.Second}
+	v, err = g.do(ctx, key, fn, &retries)
+	// if retries.num > 0 {
+	// 	fmt.Println(retries.num)
+	// }
+	return v, err
+}
+
+func (g *Group) do(ctx context.Context, key string, fn func(ctx context.Context) (interface{}, error), retries *retries) (v interface{}, err error) {
 	defer func() {
 		if errors.Cause(err) == errRetry {
+			retries.num++
 			runtime.Gosched()
-			v, err = g.Do(ctx, key, fn)
+			time.Sleep(retries.backoff)
+			retries.backoff *= 2
```

Comment:
There needs to be a limit here so know that there is no other logic that causes this to loop indefinitely.